File: backend/java/src/main/java/de/iits/elocalculatorbackend/player/service/PlayerService.java
package de.iits.elocalculatorbackend.player.service;

import de.iits.elocalculatorbackend.player.model.dto.PlayerCreateOrUpdateRequestDto;
import de.iits.elocalculatorbackend.player.model.dto.PlayerResponseDto;
import de.iits.elocalculatorbackend.player.model.entity.Player;
import de.iits.elocalculatorbackend.player.repository.PlayerRepository;
import jakarta.transaction.Transactional;
import lombok.AllArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
@AllArgsConstructor
public class PlayerService {

    private PlayerRepository repository;

    public PlayerResponseDto createPlayer(PlayerCreateOrUpdateRequestDto playerUpsertResource) {
        return new PlayerResponseDto(repository.save(playerUpsertResource.toEntity()));
    }

    public List<PlayerResponseDto> findAll() {
        return repository.findAll()
                .stream()
                .map(PlayerResponseDto::new)
                .toList();
    }

    public Optional<PlayerResponseDto> findByUsername(String username) {
        return repository.findById(username)
                .map(PlayerResponseDto::new);
    }

    @Transactional
    public PlayerResponseDto updatePlayer(PlayerCreateOrUpdateRequestDto playerUpsertResource) {
        Player playerToUpdate = repository
                .findById(playerUpsertResource.username())
                .orElseThrow(() -> new IllegalStateException(String.format("No player exists with user name " + playerUpsertResource.username())));

        update(playerToUpdate, playerUpsertResource);
        return new PlayerResponseDto(repository.save(playerToUpdate));
    }

    private void update(Player player, PlayerCreateOrUpdateRequestDto upsertResource) {
        player.setDisplayName(upsertResource.displayName());
        player.setEmail(upsertResource.email());
        player.setUsername(upsertResource.username());
    }

}

File: backend/springboot-kotlin/src/main/kotlin/de/iits/elo/player/service/PlayerService.kt
package de.iits.elo.player.service

import de.iits.elo.player.model.dto.PlayerRequestDto
import de.iits.elo.player.model.dto.PlayerResponseDto
import de.iits.elo.player.model.dto.toEntity
import de.iits.elo.player.model.entity.Player
import de.iits.elo.player.model.entity.toResponseDto
import de.iits.elo.player.repository.PlayerRepository
import org.springframework.stereotype.Service
import kotlin.jvm.optionals.getOrNull

@Service
class PlayerService(
    private val playerRepository: PlayerRepository,
) {

    fun findAll(): List<PlayerResponseDto> =
        playerRepository.findAll().map(Player::toResponseDto)

    fun findByUsername(username: String) =
        playerRepository.findById(username).getOrNull()?.toResponseDto()

    @Throws(IllegalArgumentException::class)
    fun update(player: PlayerRequestDto): PlayerResponseDto {
        val databaseObject = playerRepository.findById(player.username)
        check(databaseObject.isPresent) { "No player exists with user name ${player.username}" }
        val updateEntity = player.toEntity().copy(elo = databaseObject.get().elo)
        return playerRepository.save(updateEntity).toResponseDto()
    }

    fun create(player: PlayerRequestDto) =
        playerRepository.save(player.toEntity()).toResponseDto()
}
